File: sort-characters-by-frequency/sort-characters-by-frequency.py
class Solution:
    def frequencySort(self, s: str) -> str:
        '''
        Given a string this function creates a new string based on character frequency
        Returns the new string with most frequent characters in front
        '''
        dict = {}
        for c in s: #Collect the number of items in the string
            if c in dict:
                dict[c] += 1
            else:
                dict[c] = 1
        dict = sorted(dict.items(), key = lambda item:item[1], reverse = True) #Sort the dictionary based on key:value pair
        s = ''
        for key, number in dict: #Create the new string
            s += key * number
        return s

File: sort-characters-by-frequency/sort-characters-by-frequency.java
class Solution 
{
    public String frequencySort(String s) 
    {
        // Use a hash map to count the number of characters in the string s
        HashMap<Character, Integer> map = new HashMap<>();
        for (char ch : s.toCharArray())
            map.put(ch, map.getOrDefault(ch, 0) + 1);
        
        // Create a max heap to hold the most occurrences of a char at the top
        PriorityQueue<Map.Entry<Character, Integer>> maxHeap = new PriorityQueue<>((a,b) -> b.getValue() - a.getValue());
        maxHeap.addAll(map.entrySet());
        
        // Keep polling the heap until all elements are removed in order
        StringBuilder answer = new StringBuilder();
        while (!maxHeap.isEmpty())
        {
            Map.Entry entry = maxHeap.poll();
            for (int i = 0; i < (int)entry.getValue(); i++)
                answer.append(entry.getKey());
        }
        
        return answer.toString();
    }
}